jbmc/regression/jbmc/FloatToIntConversion/Test.java

```diff
@@ -0,0 +1,220 @@
+public class Test
+{
+  public static void floatToInt(int c, float a, float b)
+  {
+    switch(c)
+    {
+    case 0: 
+      if (a == 3.0f && b == 2.0f) {
+        int result = (int)(a / b);
+        assert result == 1;
+      }
+      break;
+    case 1: 
+      if (a == 3.0f && b == -2.0f) {
+        int result = (int)(a / b);
+        assert result == -1;
+      }
+      break;
+    case 2: 
+      if (a == 1.0f) {
+        // use constant here, -0.0f reverses cases 2 and 3.
+        int result = (int)(a / 0.0f);
+        // a/b = +inf
+        assert result == 2147483647;
+      }
+      break;
+    case 3: 
+      if (a == -1.0f) {
+        int result = (int)(a / 0.0f);
+        // a/b = -inf
+        assert result == -2147483648;
+      }
+      break;
+    case 4: 
+      if (a == 2147483646.0f && b == 1.0f) {
```

Comment:
There is quite a lot going on here and I feel like, without the comment it is easy to miss some of it.  Here is my understanding:

`2147483646` is `0x7FFFFFFE` which has 30 significant bits so `2147483646.0f` will round /in the compiler/ to `0x1p+31f` a.k.a. 2^31 which is the same as `2147483648.0f`.  The divide is effectively a no-op and so this tests the conversion to `int` but also the compiler's handling of constants.

( There are some really horrific examples like this where you can write code that will, at run-time, tell you whether the compiler correctly parsed the number as a float or whether it parsed it first as a double and cast down. )

Reply:
There are quite a lot of subtle things going on. I can comment in more detail what is happening in the test, and maybe choose constants that are not going to be rounded. I should probably also explain the effect of +1 (resp. -1) in java_bytecode_convert_method.cpp:1681, which are subject to similar rounding effects (i.e. it may not be totally obvious why this works).

Reply:
With this kind of thing, more documentation is better than less.  I know you get it and intend what happens but it should be commented so that others can learn that.